src/lib/sw_engine/tvgSwRaster.cpp

```diff
@@ -142,141 +142,104 @@ static bool _rasterTranslucentRle(SwSurface* surface, SwRleData* rle, uint32_t c
             if (rX < 0 || rX >= w || rY < 0 || rY >= h) continue;
-
-            auto dst = &surface->buffer[span->y * surface->stride + span->x + x];
-            auto index = rY * w + rX;        //TODO: need to use image's stride
-            if (dst && img && img[index]) {
-                auto src = ALPHA_BLEND(img[index], span->coverage);
-                auto invAlpha = 255 - surface->comp.alpha(src);
-                *dst = src + ALPHA_BLEND(*dst, invAlpha);
-            }
+            auto src = ALPHA_BLEND(img[rY * w + rX], span->coverage);    //TODO: need to use image's stride
+            *dst = src + ALPHA_BLEND(*dst, 255 - surface->comp.alpha(src));
+            ++dst;
         }
         ++span;
     }
     return true;
 }
 
 
-static bool _rasterTranslucentImage(SwSurface* surface, uint32_t *img, uint32_t opacity,const SwBBox& region, const Matrix* invTransform, uint32_t w, uint32_t h)
+static bool _rasterTranslucentImage(SwSurface* surface, uint32_t *img, uint32_t w, uint32_t h, uint32_t opacity, const SwBBox& region, const Matrix* invTransform)
 {
     for (auto y = region.min.y; y < region.max.y; y++) {
+        auto dst = &surface->buffer[y * surface->stride + region.min.x];
+        auto ey1 = y * invTransform->e12 + invTransform->e13;
+        auto ey2 = y * invTransform->e22 + invTransform->e23;
         for (auto x = region.min.x; x < region.max.x; x++) {
-            auto rX = static_cast<uint32_t>(roundf(x * invTransform->e11 + y * invTransform->e12 + invTransform->e13));
-            auto rY = static_cast<uint32_t>(roundf(x * invTransform->e21 + y * invTransform->e22 + invTransform->e23));
-
+            auto rX = static_cast<uint32_t>(roundf(x * invTransform->e11 + ey1));
+            auto rY = static_cast<uint32_t>(roundf(x * invTransform->e21 + ey2));
             if (rX < 0 || rX >= w || rY < 0 || rY >= h) continue;
-
-            auto dst = &surface->buffer[y * surface->stride + x];
-            auto index = rX + (rY * w); //TODO: need to use image's stride
-            if (dst && img && img[index]) {
-                auto src = ALPHA_BLEND(img[index], opacity);
-                auto invAlpha = 255 - surface->comp.alpha(src);
-                *dst = src + ALPHA_BLEND(*dst, invAlpha);
-            }
+            auto src = ALPHA_BLEND(img[rX + (rY * w)], opacity);    //TODO: need to use image's stride
+            *dst = src + ALPHA_BLEND(*dst, 255 - surface->comp.alpha(src));
+            ++dst;
         }
     }
     return true;
 }
 
 
-static bool _rasterImage(SwSurface* surface, uint32_t *img, const SwBBox& region, uint32_t w, uint32_t h)
+static bool _rasterImage(SwSurface* surface, uint32_t *img, uint32_t w, uint32_t h, const SwBBox& region)
 {
     for (auto y = region.min.y; y < region.max.y; y++) {
+        auto dst = &surface->buffer[y * surface->stride + region.min.x];
+        auto src = img + region.min.x + (y * w);    //TODO: need to use image's stride
```

Comment:
x to region.min.x is it correct?
Please check that '+ x'  is also missing here.

Reply:
not necessary.